File: dotnet/v2.6/Ads/CreateRotationGroup.cs
/*
 * Copyright 2015 Google Inc
 *
 * Licensed under the Apache License, Version 2.0(the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System;
using System.Collections.Generic;
using Google.Apis.Dfareporting.v2_6;
using Google.Apis.Dfareporting.v2_6.Data;

namespace DfaReporting.Samples {
  /// <summary>
  /// This example creates a rotation group ad in a given campaign. Start and
  /// end date for the ad must be within campaign start and end dates. To
  /// create creatives, run one of the Create*Creative.cs examples. To get
  /// available placements, run GetPlacements.cs.
  /// </summary>
  class CreateRotationGroup : SampleBase {
    /// <summary>
    /// Returns a description about the code example.
    /// </summary>
    public override string Description {
      get {
        return "This example creates a rotation group ad in a given campaign." +
            " Start and end date for the ad must be within campaign start and" +
            " end dates. To create creatives, run one of the Create*Creative.cs" +
            " examples. To get available placements, run GetPlacements.cs.\n";
      }
    }

    /// <summary>
    /// Main method, to run this code example as a standalone application.
    /// </summary>
    /// <param name="args">The command line arguments.</param>
    public static void Main(string[] args) {
      SampleBase codeExample = new CreateRotationGroup();
      Console.WriteLine(codeExample.Description);
      codeExample.Run(DfaReportingFactory.getInstance());
    }

    /// <summary>
    /// Run the code example.
    /// </summary>
    /// <param name="service">An initialized Dfa Reporting service object
    /// </param>
    public override void Run(DfareportingService service) {
      long campaignId = long.Parse(_T("INSERT_CAMPAIGN_ID_HERE"));
      long creativeId = long.Parse(_T("INSERT_CREATIVE_ID_HERE"));
      long placementId = long.Parse(_T("INSERT_PLACEMENT_ID_HERE"));
      long profileId = long.Parse(_T("INSERT_PROFILE_ID_HERE"));

      String adName = _T("INSERT_AD_NAME_HERE");

      // Retrieve the campaign.
      Campaign campaign = service.Campaigns.Get(profileId, campaignId).Execute();

      // Create a click-through URL.
      ClickThroughUrl clickThroughUrl = new ClickThroughUrl();
      clickThroughUrl.DefaultLandingPage = true;

      // Create a creative assignment.
      CreativeAssignment creativeAssignment = new CreativeAssignment();
      creativeAssignment.Active = true;
      creativeAssignment.CreativeId = creativeId;
      creativeAssignment.ClickThroughUrl = clickThroughUrl;

      // Create a placement assignment.
      PlacementAssignment placementAssignment = new PlacementAssignment();
      placementAssignment.Active = true;
      placementAssignment.PlacementId = placementId;

      // Create a creative rotation.
      CreativeRotation creativeRotation = new CreativeRotation();
      creativeRotation.CreativeAssignments = new List<CreativeAssignment>() {
          creativeAssignment
      };

      // Create a delivery schedule.
      DeliverySchedule deliverySchedule = new DeliverySchedule();
      deliverySchedule.ImpressionRatio = 1;
      deliverySchedule.Priority = "AD_PRIORITY_01";

      DateTime startDate = DateTime.Now;
      DateTime endDate = Convert.ToDateTime(campaign.EndDate);

      // Create a rotation group.
      Ad rotationGroup = new Ad();
      rotationGroup.Active = true;
      rotationGroup.CampaignId = campaignId;
      rotationGroup.CreativeRotation = creativeRotation;
      rotationGroup.DeliverySchedule = deliverySchedule;
      rotationGroup.StartTime = startDate;
      rotationGroup.EndTime = endDate;
      rotationGroup.Name = adName;
      rotationGroup.PlacementAssignments = new List<PlacementAssignment>() {
          placementAssignment
      };
      rotationGroup.Type = "AD_SERVING_STANDARD_AD";

      // Insert the rotation group.
      Ad result = service.Ads.Insert(rotationGroup, profileId).Execute();

      // Display the new ad ID.
      Console.WriteLine("Ad with ID {0} was created.", result.Id);
    }
  }
}

File: ruby/v3.2/ads/create_rotation_group.rb
#!/usr/bin/env ruby

#
# Copyright:: Copyright 2016, Google Inc. All Rights Reserved.
#
# License:: Licensed under the Apache License, Version 2.0 (the "License");
#           you may not use this file except in compliance with the License.
#           You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
#           Unless required by applicable law or agreed to in writing, software
#           distributed under the License is distributed on an "AS IS" BASIS,
#           WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
#           implied.
#           See the License for the specific language governing permissions and
#           limitations under the License.
#
# This example creates a rotation group ad in a given campaign.

require_relative '../dfareporting_utils'

def create_rotation_group(profile_id, campaign_id, placement_id, creative_id)
  # Authenticate and initialize API service.
  service = DfareportingUtils.get_service

  # Retrieve the campaign (to get end date).
  campaign = service.get_campaign(profile_id, campaign_id)

  # Construct creative assignment.
  creative_assignment =
    DfareportingUtils::API_NAMESPACE::CreativeAssignment.new(
      active: true,
      creative_id: creative_id,
      click_through_url: DfareportingUtils::API_NAMESPACE::ClickThroughUrl.new(
        default_landing_page: true
      )
    )

  # Construct placement assignment.
  placement_assignment =
    DfareportingUtils::API_NAMESPACE::PlacementAssignment.new(
      active: true,
      placement_id: placement_id
    )

  # Construct creative rotation.
  creative_rotation = DfareportingUtils::API_NAMESPACE::CreativeRotation.new(
    creative_assignments: [creative_assignment],
    type: 'CREATIVE_ROTATION_TYPE_RANDOM',
    weight_calculation_strategy: 'WEIGHT_STRATEGY_OPTIMIZED'
  )

  # Construct delivery schedule.
  delivery_schedule = DfareportingUtils::API_NAMESPACE::DeliverySchedule.new(
    impression_ratio: 1,
    priority: 'AD_PRIORITY_01'
  )

  # Construct and save ad.
  ad = DfareportingUtils::API_NAMESPACE::Ad.new(
    active: true,
    campaign_id: campaign_id,
    creative_rotation: creative_rotation,
    delivery_schedule: delivery_schedule,
    end_time: format('%sT00:00:00Z', campaign.end_date),
    name: 'Example Rotation Group',
    placement_assignments: [placement_assignment],
    start_time: format('%sT23:59:59Z', Time.now.strftime('%Y-%m-%d')),
    type: 'AD_SERVING_STANDARD_AD'
  )

  result = service.insert_ad(profile_id, ad)

  puts format('Created rotation group with ID %d and name "%s".', result.id, result.name)
end

if $PROGRAM_NAME == __FILE__
  # Retrieve command line arguments.
  args = DfareportingUtils.get_arguments(ARGV, :profile_id, :campaign_id,
    :placement_id, :creative_id)

  create_rotation_group(args[:profile_id], args[:campaign_id],
    args[:placement_id], args[:creative_id])
end
